src/compiler/crystal/semantic/call_error.cr

```diff
@@ -277,7 +277,15 @@ class Crystal::Call
     index_or_name = indexes_or_names_in_all_overloads.first
 
     mismatches = argument_type_mismatches.select(&.index_or_name.==(index_or_name))
-    expected_types = mismatches.map(&.expected_type).uniq.sort_by(&.to_s)
+    expected_types = mismatches.map(&.expected_type).uniq!
+    expected_types = expected_types.flat_map do |expected_type|
+      if expected_type.is_a?(UnionType)
+        expected_type.union_types
```

Comment:
Probably a corner case, but if you have a `((Int | Char) | String)`, wouldn't this just strip the outermost union?

Reply:
A union type can never have another union type inside it. They are always flatten when forming a `UnionType`.